Querying for a tile that is outside the zoom limits results in 500. I think it should return a 404 instead.

``` http
> http :8090/nlcd-agg-tiles/13/2355/3026.png
HTTP/1.1 500 Internal Server Error
Content-Length: 35
Content-Type: text/plain; charset=UTF-8
Date: Wed, 07 Sep 2016 21:36:08 GMT
Server: spray-can/1.3.3

There was an internal server error.
```

Here's what the service output looks like:

```
[ERROR] [09/07/2016 21:36:08.039] [ForkJoinPool-4-worker-3] [akka://usace-programanalysis-geop/user/usace-programanalysis] Error during processing of request HttpRequest(GET,http://localhost:8090/nlcd-agg-tiles/13/2355/3026.png,List(Host: localhost:8090, User-Agent: HTTPie/0.9.4, Connection: keep-alive, Accept-Encoding: gzip, deflate),Empty,HTTP/1.1)
geotrellis.spark.io.package$AttributeNotFoundError: Attribute metadata not found for layer Layer(name = "nlcd-zoomed", zoom = 13)
    at geotrellis.spark.io.s3.S3AttributeStore.read(S3AttributeStore.scala:55)
    at geotrellis.spark.io.AttributeCaching$$anonfun$cacheRead$1.apply(AttributeCaching.scala:11)
    at scala.collection.mutable.MapLike$class.getOrElseUpdate(MapLike.scala:189)
    at scala.collection.mutable.AbstractMap.getOrElseUpdate(Map.scala:91)
    at geotrellis.spark.io.AttributeCaching$class.cacheRead(AttributeCaching.scala:11)
    at geotrellis.spark.io.s3.S3AttributeStore.cacheRead(S3AttributeStore.scala:20)
    at geotrellis.spark.io.BlobLayerAttributeStore$class.readHeader(AttributeStore.scala:60)
    at geotrellis.spark.io.s3.S3AttributeStore.readHeader(S3AttributeStore.scala:20)
    at geotrellis.spark.io.s3.S3ValueReader$$anon$1.<init>(S3ValueReader.scala:24)
    at geotrellis.spark.io.s3.S3ValueReader.reader(S3ValueReader.scala:23)
    at com.azavea.usace.programanalysis.geop.LayerReader$.catalog(LayerReader.scala:51)
    at com.azavea.usace.programanalysis.geop.LayerReader$.catalog(LayerReader.scala:41)
    at com.azavea.usace.programanalysis.geop.LayerReader$.apply(LayerReader.scala:30)
    at com.azavea.usace.programanalysis.geop.GeopServiceActor$$anonfun$tilesHandler$1$$anonfun$apply$5$$anonfun$apply$6.apply(GeopServiceActor.scala:91)
    at com.azavea.usace.programanalysis.geop.GeopServiceActor$$anonfun$tilesHandler$1$$anonfun$apply$5$$anonfun$apply$6.apply(GeopServiceActor.scala:90)
    at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
    at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
    at scala.concurrent.impl.ExecutionContextImpl$$anon$3.exec(ExecutionContextImpl.scala:107)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: com.amazonaws.services.s3.model.AmazonS3Exception: The specified key does not exist. (Service: Amazon S3; Status Code: 404; Error Code: NoSuchKey; Request ID: 70316D50CDBB614B), S3 Extended Request ID: c5t0sUD5cFQ43KoxogUsisvJRlp8eGwtGooYv20bmiAcRYl+E28jmzv0Qxd+UI4X81jKqygo7e8=
    at com.amazonaws.http.AmazonHttpClient.handleErrorResponse(AmazonHttpClient.java:1127)
    at com.amazonaws.http.AmazonHttpClient.executeOneRequest(AmazonHttpClient.java:743)
    at com.amazonaws.http.AmazonHttpClient.executeHelper(AmazonHttpClient.java:462)
    at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:297)
    at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:3672)
    at com.amazonaws.services.s3.AmazonS3Client.getObject(AmazonS3Client.java:1160)
    at geotrellis.spark.io.s3.AmazonS3Client.getObject(S3Client.scala:149)
    at geotrellis.spark.io.s3.S3Client$class.getObject(S3Client.scala:48)
    at geotrellis.spark.io.s3.AmazonS3Client.getObject(S3Client.scala:121)
    at geotrellis.spark.io.s3.S3AttributeStore.geotrellis$spark$io$s3$S3AttributeStore$$readKey(S3AttributeStore.scala:39)
    at geotrellis.spark.io.s3.S3AttributeStore.read(S3AttributeStore.scala:52)
    ... 21 more
```

This is the interesting part:

```
Caused by: com.amazonaws.services.s3.model.AmazonS3Exception: The specified key does not exist. (Service: Amazon S3; Status Code: 404; Error Code: NoSuchKey; Request ID: 70316D50CDBB614B), S3 Extended Request ID: c5t0sUD5cFQ43KoxogUsisvJRlp8eGwtGooYv20bmiAcRYl+E28jmzv0Qxd+UI4X81jKqygo7e8=
```

So S3 is returning a 404. We should pass that on, instead of returning a 500.

--- a/geop/src/main/scala/com/azavea/usace/programanalysis/geop/ClippedLayers.scala
+++ b/geop/src/main/scala/com/azavea/usace/programanalysis/geop/ClippedLayers.scala
@@ -3,7 +3,7 @@ package com.azavea.usace.programanalysis.geop
 import geotrellis.raster.Tile
 import geotrellis.spark.io._
 import geotrellis.spark.{SpatialKey, TileLayerMetadata, _}
-import geotrellis.spark.io.s3.{S3AttributeStore, S3LayerReader, S3ValueReader}
+import geotrellis.spark.io.s3.{S3AttributeStore, S3LayerReader}
 import geotrellis.vector.{Extent, MultiPolygon}
 
 import org.apache.spark.SparkContext
@@ -26,25 +26,11 @@ object ClippedLayers {
     sc: SparkContext
   ): Seq[TileLayerRDD[SpatialKey]] = {
     val extent = multiPolygon.envelope
-    println(multiPolygon.toString)
-    println(extent.toString)
     val rasterLayers = rasterLayerIds.map(rasterLayerId =>
-      queryAndCropLayer(catalog(sc), rasterLayerId, Some(extent))
+      queryAndCropLayer(catalog(sc), rasterLayerId, extent)
     )
 
     rasterLayers
-  }
-
-  def apply(
-    zoom: Int,
-    x: Int,
-    y: Int,
-    sc: SparkContext
-  ): Tile = {
-    val sKey: SpatialKey = SpatialKey(x, y)
-    val rdr = catalog(sc, zoom)
-
-    rdr.read(sKey)
   }
 
   /**
@@ -59,18 +45,11 @@ object ClippedLayers {
   def queryAndCropLayer(
     catalog: S3LayerReader,
     layerId: LayerId,
-    extent: Option[Extent]
+    extent: Extent
   ): TileLayerRDD[SpatialKey] = {
-    extent match {
-      case Some(extent) =>
-        catalog.query[SpatialKey, Tile, TileLayerMetadata[SpatialKey]](layerId)
-          .where(Intersects(extent))
-          .result
-      case None =>
-        catalog.query[SpatialKey, Tile, TileLayerMetadata[SpatialKey]](layerId)
-          .where(Intersects(KeyBounds(SpatialKey(10, 10), SpatialKey(10, 10))))
-          .result
-    }
+    catalog.query[SpatialKey, Tile, TileLayerMetadata[SpatialKey]](layerId)
+      .where(Intersects(extent))
+      .result
   }
 
   /**
@@ -80,9 +59,6 @@ object ClippedLayers {
     */
   def catalog(sc: SparkContext): S3LayerReader =
     catalog("azavea-datahub", "catalog")(sc)
-  
-  def catalog(sc: SparkContext, zoom: Int): Reader[SpatialKey, Tile] =
-    catalog("azavea-datahub", "catalog", zoom)(sc)
 
   /**
     * Given an S3 bucket name and root path, returns a catalog of layers
@@ -100,20 +76,6 @@ object ClippedLayers {
   )(implicit sc: SparkContext): S3LayerReader = {
     val attributeStore = new S3AttributeStore(bucket, rootPath)
     val reader = new S3LayerReader(attributeStore)
-
-    reader
-  }
-
-  
-  def catalog(
-    bucket: String,
-    rootPath: String,
-    zoom: Int
-  )(implicit sc: SparkContext): Reader[SpatialKey, Tile] = {
-    val attributeStore = new S3AttributeStore(bucket, rootPath)
-    val vr = new S3ValueReader(attributeStore)
-    val layer = LayerId("nlcd-zoomed", zoom)
-    val reader: Reader[SpatialKey, Tile] = vr.reader(layer)
 
     reader
   }

--- a/geop/src/main/scala/com/azavea/usace/programanalysis/geop/GeopServiceActor.scala
+++ b/geop/src/main/scala/com/azavea/usace/programanalysis/geop/GeopServiceActor.scala
@@ -12,12 +12,11 @@ import spray.http.HttpMethods.{DELETE, GET, OPTIONS, POST}
 import spray.json.{JsNumber, JsObject}
 import spray.routing.{Directive0, HttpService, RejectionHandler}
 
-import geotrellis.raster.Tile
-import geotrellis.spark.io.{TileNotFoundError}
 import geotrellis.proj4.{LatLng, ConusAlbers}
-import geotrellis.raster.{IntConstantNoDataCellType}
-import geotrellis.raster.render.{IntColorMap, ClassBoundaryType, Exact}
-import geotrellis.raster.render.ColorMap.{Options}
+import geotrellis.raster.{IntConstantNoDataCellType, Tile}
+import geotrellis.raster.render.{IntColorMap, Exact}
+import geotrellis.raster.render.ColorMap.Options
+import geotrellis.spark.io.{TileNotFoundError, AttributeNotFoundError}
 
 
 class GeopServiceActor(sc: SparkContext) extends Actor with HttpService {
@@ -68,15 +67,15 @@ class GeopServiceActor(sc: SparkContext) extends Actor with HttpService {
         complete {
           future {
             args.multiPolygons.map(m => {
-                val multiPolygon = m.reproject(LatLng, ConusAlbers)
-                val rasterLayers = ClippedLayers(args.rasters, multiPolygon, sc)
-                val rasterGroupedCount = RasterGroupedCount(rasterLayers, multiPolygon)
+              val multiPolygon = m.reproject(LatLng, ConusAlbers)
+              val rasterLayers = ClippedLayers(args.rasters, multiPolygon, sc)
+              val rasterGroupedCount = RasterGroupedCount(rasterLayers, multiPolygon)
 
-                JsObject(
-                  rasterGroupedCount
-                    .map { case (keys, count) =>
-                      keys.mkString(",") -> JsNumber(count)
-                    })
+              JsObject(
+                rasterGroupedCount
+                  .map { case (keys, count) =>
+                    keys.mkString(",") -> JsNumber(count)
+                  })
             }).toVector
           }
         }
@@ -90,9 +89,9 @@ class GeopServiceActor(sc: SparkContext) extends Actor with HttpService {
           future {
             val result: Option[Tile] =
               try {
-                Some(ClippedLayers(zoom, x, y, sc))
+                Some(LayerReader(zoom, x, y, sc))
               } catch {
-                case _: TileNotFoundError => None
+                case _: AttributeNotFoundError | _: TileNotFoundError => None
               }
 
             result.map { tile =>

--- a/geop/src/main/scala/com/azavea/usace/programanalysis/geop/JsonProtocol.scala
+++ b/geop/src/main/scala/com/azavea/usace/programanalysis/geop/JsonProtocol.scala
@@ -32,7 +32,7 @@ object JsonProtocol extends SprayJsonSupport with GeoJsonSupport {
     }
 
     def read(value: JsValue) = {
-      value.asJsObject.getFields("zoom", "rasters", "multiPolygons", "lat", "lng") match {
+      value.asJsObject.getFields("zoom", "rasters", "multiPolygons") match {
         case Seq(JsNumber(zoom), JsArray(rasters), JsArray(multiPolygons)) =>
           new CountArgs(
             rasters.map { r => LayerId(r.convertTo[String], zoom.toInt) },

--- a/geop/src/main/scala/com/azavea/usace/programanalysis/geop/LayerReader.scala
+++ b/geop/src/main/scala/com/azavea/usace/programanalysis/geop/LayerReader.scala
@@ -1,0 +1,53 @@
+package com.azavea.usace.programanalysis.geop
+
+import geotrellis.raster.Tile
+import geotrellis.spark.io._
+import geotrellis.spark.{SpatialKey, LayerId}
+import geotrellis.spark.io.s3.{S3AttributeStore, S3ValueReader}
+
+import org.apache.spark.SparkContext
+
+object LayerReader {
+  /**
+    * Given values for zoom, x and y, and a spark context, returns
+    * the corresponding [[Tile]] object.
+    *
+    * @param    zoom  [[Int]] value of zoom level
+    * @param    x     [[Int]] x-value of the spatial key
+    * @param    y     [[Int]] y-value of the spatial key
+    * @param    sc    [[SparkContext]] for the [[S3ValueReader]]
+    * @return         [[Tile]] object
+    */
+  def apply(
+    zoom: Int,
+    x: Int,
+    y: Int,
+    sc: SparkContext
+  ): Tile = {
+    val sKey = SpatialKey(x, y)
+    val rdr = catalog(sc, zoom)
+
+    rdr.read(sKey)
+  }
+
+  /**
+    * Given a spark context, returns the correct catalog. This configures the
+    * next method with defaults.
+    *
+    */
+  def catalog(sc: SparkContext, zoom: Int): Reader[SpatialKey, Tile] =
+    catalog("azavea-datahub", "catalog", zoom)(sc)
+
+  def catalog(
+    bucket: String,
+    rootPath: String,
+    zoom: Int
+  )(implicit sc: SparkContext): Reader[SpatialKey, Tile] = {
+    val attributeStore = new S3AttributeStore(bucket, rootPath)
+    val vr = new S3ValueReader(attributeStore)
+    val layer = LayerId("nlcd-zoomed", zoom)
+    val reader: Reader[SpatialKey, Tile] = vr.reader(layer)
+
+    reader
+  }
+}

